### ISSUE TITLE
ClassInfo.getClassDependencies doesn't look into method body.

### ISSUE THREAD

**Issue by PYangDizzle** [classgraph/classgraph#413]
According to the ClassInfo wiki, ClassGraph is capable of inspecting the method implementation to find dependencies.

> ClassInfo#getClassDependencies() returns a ClassInfoList for all the classes a given class depends upon, by looking for class references in superclasses, interfaces, methods, fields, annotations, **local variables, intermediate values within a method's code**, concrete type parameters, etc.

However, `getClassDependencies` doesn't do that.

I turned on verbose and noticed that _temp.Source_ is recognized at one point.

> 2020-04-03T17:52:34.548-0500	ClassGraph	------------ Found class temp.RegressionTarget
2020-04-03T17:52:34.548-0500	ClassGraph	-------------- Superclass: java.lang.Object
2020-04-03T17:52:34.548-0500	ClassGraph	-------------- Method:  <init>()
2020-04-03T17:52:34.549-0500	ClassGraph	-------------- Method: public static void doSomething()
2020-04-03T17:52:34.549-0500	ClassGraph	-------------- Referenced class names: java.lang.Object, temp.RegressionTarget, temp.Source

I further debugged and noticed that ClassInfo has 2 fields that are semantically tightly coupled but programmatically unrelated. 
```
    /**
     * Names of classes referenced by this class in class refs and type signatures in the constant pool of the
     * classfile.
     */
    private Set<String> referencedClassNames;

    /** A list of ClassInfo objects for classes referenced by this class. */
    private ClassInfoList referencedClasses;
```
`referencedClasses` is generated from scratch even if `referencedClassNames` is already populated.
`referencedClasses` doesn't have `temp.Source`, but `referencedClassNames` does.

I think `io.github.classgraph.ScanResult.indexResourcesAndClassInfo()` should utilize `referencedClassNames`.

I attached a full log from `verbose()` and a failing JUnit 5 test case because of this bug.

[log.txt](https://github.com/classgraph/classgraph/files/4430245/log.txt)
[ReferencedClassTestCase.txt](https://github.com/classgraph/classgraph/files/4430253/ReferencedClassTestCase.txt)

**Comment by lukehutch** [comment]
Hmm.. the unit test passes for me. I checked it into git master in the above commit -- can you please see if you can get it to break? I'll take a look at the other code you referenced.

**Comment by PYangDizzle** [comment]
@lukehutch 
I was able to make it fail with 2 different ways.

1. I removed the whitelisting to use the default whitelisting (which should be all packages except system modules) and this cause the test to fail.
```
ScanResult sr = new ClassGraph()
//        		.whitelistPackages(getClass().getPackage().getName())
                .enableInterClassDependencies().whitelistClasses(typeName).scan();
```

2. I moved the 2 target classes `RegressionTarget` and `Source` into a different package and the test failed.

**Comment by lukehutch** [comment]
Ah, OK, this is actually working as intended:

1. The code in `ScanResult.indexResourcesAndClassInfo()` uses `referencedClassNames` to find or create `ClassInfo` objects for each class, and these are stored in `referencedClasses`.

2. The reason you weren't seeing results is that if nothing is whitelisted, everything is. But as soon as you whitelist something specific, only the specifically-whitelisted things are treated as whitelisted. Maybe the documentation needs to make that more clear.

In this case, since you had whitelisted a class by name, the package containing the class is *not* whitelisted (i.e. other classes in that package are not scanned) unless you specifically whitelist the package itself. And that's because there would be no point whitelisting the class if the package it is contained in is also whitelisted.

I know that the logic that "(only) if nothing is whitelisted, everything is" is a little bit weird. But that gives the most natural behavior in the widest array of circumstances.

Please confirm that this answer resolves your concern, and if there's still a bug, I'll look again.

**Comment by lukehutch** [comment]
So this is what is currently in the docs:

> 💡 If no packages, classes or paths are whitelisted, then all packages or paths are scanned. Note that package and path whitelisting/blacklisting work the same internally, they just accept different separator characters ('.' for packages, '/' for paths), meaning that if you whitelist/blacklist a path, the corresponding package is whitelisted/blacklisted, and vice versa.

I added this:

> 💡 If nothing is whitelisted, everything is. But if you whitelist a specific package or class, then other packages or classes will not be scanned unless they too are whitelisted. Therefore, if you whitelist a specific class, other classes in the same package will not be scanned unless the package itself is also whitelisted -- and if the package is whitelisted, you don't need to whitelist the specific class, because it resides within the whitelisted package.

**Comment by lukehutch** [comment]
By the way this failed for related reasons:

> 2. I moved the 2 target classes `RegressionTarget` and `Source` into a different package and the test failed.

This is because with package whitelisting in place, the other package you moved the classes into is not scanned.

**Comment by lukehutch** [comment]
So actually I'm pretty confident this is not a bug, and I'll close it -- but if you do identify a bug in spite of the above information, please let me know and I'll reopen this.

**Comment by lukehutch** [comment]
PS @PYangDizzle Thanks for the report! And for digging deeper to understand what was going on.

**Comment by PYangDizzle** [comment]
@lukehutch
I enabled externalClasses and it returned correctly. Thanks for the clarification.

I was so sure that whitelisting is unrelated because I added it only to minimize the logs for debugging the error. I realized that I downloaded an old version by mistake and got an updated one but that was after I added whitelisting.

So, the order or event is
1. Test with 4.8.9 -> failed because it's an old version
2. Enabled whitelisting while debugging -> still an old version
3. Test with 4.8.67 -> failed because whitelisting is enabled

The mistake was because of how sonatype orders the folders. 4.8.9 is at the end...
![image](https://user-images.githubusercontent.com/15711504/78454808-ff92c180-765f-11ea-9e25-104d6bac5dc3.png)
